### PR TITLE
✨Add a library to parse container image name and other fields

### LINKED ISSUE #2850: Provide a library to work with container images
We should have a package under `util/` to deal with container images, more specifically:

- We can copy this function from k/k https://github.com/kubernetes/kubernetes/blob/master/pkg/util/parsers/parsers.go#L36-L59 to add basic validation.
- Our parsing function will have to do some extra parsing to split the name of the repository (last part after splitting by `/`) and the base repository.
- Create a struct similar to
```go
type Image struct {
   Repository
   Name
   Tag
   Digest
}
```

As an example, let's take the string `"k8s.gcr.io/dev/coredns:1.6.2"`, when parsed it should return
```go
Image{
  Repository: "k8s.gcr.io/dev"
  Name: "coredns"
  Tag: "1.6.2"
  Digest: ""
}
```

/kind feature
/milestone v0.3.x

### PR DESCRIPTION
**Pull request by nader-ziada**
**What this PR does / why we need it**:
 - Library to parse image name. and other information  
 - Function will split the name of the repository in a seperate field

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2850

### CODE REVIEW

**Comment by nader-ziada** [comment]
@vincepri there is only a couple of places using the reference library, I can include them in this PR

File: util/container/image_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			image, err := ParseImageName(tc.input)
+			if tc.wantError {
+				g.Expect(err).To(HaveOccurred())
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+			g.Expect(image.Repository).To(Equal(tc.repo))
+			g.Expect(image.Name).To(Equal(tc.imageName))
+			g.Expect(image.Tag).To(Equal(tc.tag))
+			g.Expect(image.Digest).To(Equal(tc.digest))
```
Comment (benmoss) [inline util/container/image_test.go:134]: Shouldn't this test be able to pass?

```
g.Expect(image.String()).To(Equal(tc.input))
```

File: util/util.go [hunk 2]
```diff
@@ -149,25 +150,25 @@ func ModifyImageTag(imageName, tagName string) (string, error) {
 
 // ModifyImageRepository takes an imageName (e.g., repository/image:tag), and returns an image name with updated repository
 func ModifyImageRepository(imageName, repositoryName string) (string, error) {
```
Comment (benmoss) [inline util/util.go:152]: Shouldn't we be able to get rid of these `ModifyImage` helpers entirely? I would think you can just do like 

```
image, _ := container.ParseImageName(imageName)
image.Repository = repositoryName
image.String()
```

**Comment by nader-ziada** [comment]
/hold

making some more change to this PR to apply the change to other places in the code using these functions

File: util/container/image.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+func ParseImageName(image string) (Image, error) {
+	named, err := dockerref.ParseNamed(image)
+	if err != nil {
+		return Image{}, fmt.Errorf("couldn't parse image name: %v", err)
+	}
+
+	var repo, tag, digest string
+	_, nameOnly := path.Split(reference.Path(named))
+	if nameOnly != "" {
+		repo = named.Name()[:len(named.Name())-len(nameOnly)-1]
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply util/container/image.go]: added a comment and tried to simplify

File: util/container/image.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	"fmt"
+	"path"
+
+	//  Import the crypto sha256 algorithm for the docker image parser to work
+	_ "crypto/sha256"
+	//  Import the crypto/sha512 algorithm for the docker image parser to work with 384 and 512 sha hashes
+	_ "crypto/sha512"
+
+	"github.com/docker/distribution/reference"
+	dockerref "github.com/docker/distribution/reference"
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply util/container/image.go]: fixed it

File: util/container/image.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+const (
+	// DefaultImageTag is the default tag for docker image.
+	DefaultImageTag = "latest"
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply util/container/image.go]: removed it

File: util/util.go (hunk 2 above)
Comment (nader-ziada) [reply util/util.go]: if we remove all the calls to the helpers, we have to duplicate a bunch of validation,  but tried to make the use of the `docker reference` helpers in one file only

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 6]
```diff
@@ -38,6 +38,7 @@ import (
 	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
 	"sigs.k8s.io/cluster-api/util"
 	"sigs.k8s.io/cluster-api/util/certs"
+	imageutil "sigs.k8s.io/cluster-api/util/container"
```
Comment (vincepri) [inline controlplane/kubeadm/internal/workload_cluster.go:41]: Why not use `container` like everywhere else?

File: util/util.go [hunk 7]
```diff
@@ -125,56 +122,6 @@ func Ordinalize(n int) string {
 	return fmt.Sprintf("%d%s", n, m[an%10])
 }
 
-// ModifyImageTag takes an imageName (e.g., repository/image:tag), and returns an image name with updated tag
-func ModifyImageTag(imageName, tagName string) (string, error) {
```
Comment (vincepri) [inline util/util.go:129]: If we move these functions out of the `util` package, that's an API breaking change for clients, I'd suggest to avoid this in v0.3.x, copy the functions and call the new ones in the new package from here. 

Add a comment on top of it that mentions that these are deprecated and have been moved.

File: controlplane/kubeadm/internal/workload_cluster.go (hunk 6 above)
Comment (nader-ziada) [reply controlplane/kubeadm/internal/workload_cluster.go:41]: it was conflicting with a actual container in the same func

File: util/util.go (hunk 7 above)
Comment (nader-ziada) [reply util/util.go:129]: good point, didn't realize clients could be using that, will put them back with a deprecation comment

File: controlplane/kubeadm/internal/workload_cluster.go (hunk 6 above)
Comment (vincepri) [reply controlplane/kubeadm/internal/workload_cluster.go:41]: Should probably be `containerutil`, wdyt?

Comment (nader-ziada) [reply controlplane/kubeadm/internal/workload_cluster.go:41]: sure, will use `containerutil `

File: util/container/image_test.go (hunk 1 above)
Comment (nader-ziada) [reply util/container/image_test.go]: nice catch, will add

File: util/container/image.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	//  Import the crypto/sha512 algorithm for the docker image parser to work with 384 and 512 sha hashes
+	_ "crypto/sha512"
+
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+)
+
+const (
+	// NameTotalLengthMax is the maximum total number of characters in a repository name.
+	NameTotalLengthMax = 255
```
Comment (benmoss) [inline util/container/image.go:35]: Haha wait did you delete one unused constant and add another 😄

File: util/container/image.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+	return reference.FamiliarString(reference.TagNameOnly(namedTagged)), nil
+}
+
+// ImageTagIsValid ensures that a given image tag is compliant with the OCI spec
+func ImageTagIsValid(tagName string) bool {
+	return !ociTagAllowedChars.MatchString(tagName)
+}
+
+// SemVerToOCIImageTag is a helper function that replaces all
```
Comment (benmoss) [inline util/container/image.go:138]:
```suggestion
// SemverToOCIImageTag is a helper function that replaces all
```

File: util/container/image.go (hunk 8 above)
Comment (nader-ziada) [reply util/container/image.go:35]: yeah, will remove that one too :D

File: util/util_test.go [hunk 10]
```diff
@@ -23,9 +23,9 @@ import (
 	"testing"
 
 	"github.com/blang/semver"
+	"github.com/docker/distribution/reference"
 	. "github.com/onsi/gomega"
```
Comment (vincepri) [inline util/util_test.go:28]: Remove this space?

File: cmd/clusterctl/client/config/imagemeta_client.go [hunk 11]
```diff
@@ -119,53 +118,20 @@ func (i *imageMeta) Union(other *imageMeta) {
 
 // ApplyToImage changes an image name applying the transformations defined in the current imageMeta.
 func (i *imageMeta) ApplyToImage(image string) (string, error) {
```
Comment (vincepri) [inline cmd/clusterctl/client/config/imagemeta_client.go:120]: cc @fabriziopandini @wfernandes 

to make sure we have tests and the logic looks good

File: util/util.go [hunk 12]
```diff
@@ -191,13 +158,14 @@ func GetMachinesForCluster(ctx context.Context, c client.Client, cluster *cluste
...
+// SemverToOCIImageTag is a helper function that replaces all
 // non-allowed symbols in tag strings with underscores.
 // Image tag can only contain lowercase and uppercase letters, digits,
 // underscores, periods and dashes.
 // Current usage is for CI images where all of symbols except '+' are valid,
 // but function is for generic usage where input can't be always pre-validated.
 // Taken from k8s.io/cmd/kubeadm/app/util
+// Deprecated: Please use the functions in util/container
 func SemverToOCIImageTag(version string) string {
 	return ociTagAllowedChars.ReplaceAllString(version, "_")
```
Comment (vincepri) [inline util/util.go:170]: This should probably also call the similar function in the `util/container` package

File: util/util.go [hunk 13]
```diff
@@ -125,52 +124,20 @@ func Ordinalize(n int) string {
...
+// ModifyImageTag takes an imageName (e.g., repository/image:tag), and returns an image name with updated tag
+// Deprecated: Please use the functions in util/container
+func ModifyImageTag(imageName, tagName string) (string, error) {
+	return container.ModifyImageTag(imageName, tagName)
 }
 
 // ImageTagIsValid ensures that a given image tag is compliant with the OCI spec
+// Deprecated: Please use the functions in util/container
 func ImageTagIsValid(tagName string) bool {
 	return !ociTagAllowedChars.MatchString(tagName)
```
Comment (vincepri) [inline util/util.go:142]: Same here, call the one in `container` package

Comment (vincepri) [reply util/util.go:142]: We can also remove the local variable afterwards

File: cmd/clusterctl/client/config/imagemeta_client.go (hunk 11 above)
Comment (wfernandes) [reply cmd/clusterctl/client/config/imagemeta_client.go:120]: In `imagemeta_client_test.go` there were no specific unit tests for `ApplyToImage` but there were tests for `AlterImage` which calls out to `ApplyImage`. That being said I see that a lot of tests were added for `ImageFromString`. 👍

File: util/container/image.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+	Repository string
+	Name       string
+	Tag        string
+	Digest     string
+}
+
+// ImageFromString parses a docker image string into three parts: repo, tag and digest.
+// If both tag and digest are empty, a default image tag will be returned.
+func ImageFromString(image string) (Image, error) {
+	named, err := reference.ParseNamed(image)
```
Comment (wfernandes) [inline util/container/image.go:48]: In the previous implementation, we used `reference.ParseNormalizedNamed(image)`. Is there some essential difference between `ParseNormalizedNamed` and `ParseName`?

File: util/container/image_test.go [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+		},
+		{
+			name:      "input with path and name",
+			input:     "example.com/user/repo",
+			repo:      "example.com/user",
+			imageName: "repo",
+			wantError: false,
+		},
+		{
+			name:      "input with path, anme and tag",
```
Comment (wfernandes) [inline util/container/image_test.go:76]: nit:
```suggestion
			name:      "input with path, name and tag",
```

File: util/container/image.go [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+// Image type represents the container image details
+type Image struct {
+	Repository string
+	Name       string
+	Tag        string
+	Digest     string
+}
+
+// ImageFromString parses a docker image string into three parts: repo, tag and digest.
+// If both tag and digest are empty, a default image tag will be returned.
```
Comment (wfernandes) [inline util/container/image.go:46]: Is this correct? What is the default image tag being returned? For the test case below `input with name only` we aren't asserting on any tag.

File: util/container/image_test.go [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+			tag:       "tag",
+			wantError: false,
+		},
+		{
+			name:      "input with path, name, tag and digest",
+			input:     "example.com/user/repo@sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+			repo:      "example.com/user",
+			imageName: "repo",
+			digest:    "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+			wantError: false,
```
Comment (wfernandes) [inline util/container/image_test.go:89]: Can we have an image with tag and digest specified? This test doesn't have a tag specified.

File: util/container/image.go [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+		if err != nil {
+			return "", errors.Wrap(err, "failed to parse image tag")
+		}
+		return reference.FamiliarString(retagged), nil
+	}
+	return "", errors.New("image must be tagged")
+}
+
+// ModifyImageTag takes an imageName (e.g., repository/image:tag), and returns an image name with updated tag
+func ModifyImageTag(imageName, tagName string) (string, error) {
```
Comment (wfernandes) [inline util/container/image.go:107]: Out of curiosity why are these functions called `Modify*`? Should they just be called `ImageTag` and `ImageRepository` since they return those values respectively?

Comment (wfernandes) [reply util/container/image.go:107]: Nevermind I saw the tests below which cleared this up.

File: util/container/image_test.go [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+				g.Expect(image.Digest).To(Equal(tc.digest))
+				g.Expect(image.String()).To(Equal(tc.input))
+			}
+		})
+	}
+}
+
+func TestModifyImageRepository(t *testing.T) {
+	const testRepository = "example.com/new"
+	g := NewGomegaWithT(t)
```
Comment (wfernandes) [inline util/container/image_test.go:144]: nit: Can we change this to `NewWithT` since `NewGomegaWithT` is deprecated?

File: util/container/image.go (hunk 18 above)
Comment (nader-ziada) [reply util/container/image.go:107]: I think just saying `ImageTag` could imply the func is just returning a value, buts its actually modifying the value from imageName with the new tag

File: util/container/image_test.go [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+				g.Expect(image.Name).To(Equal(tc.imageName))
+				g.Expect(image.Tag).To(Equal(tc.tag))
+				g.Expect(image.Digest).To(Equal(tc.digest))
+				g.Expect(image.String()).To(Equal(tc.input))
+			}
+		})
+	}
+}
+
+func TestModifyImageRepository(t *testing.T) {
```
Comment (wfernandes) [inline util/container/image_test.go:142]: Can we convert this into a table test since we are here? Or should we do that in a separate PR?

File: util/container/image_test.go [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+		g.Expect(err).To(MatchError(ContainSubstring("image must be tagged")))
+	})
+	t.Run("errors if the image name is not valid", func(t *testing.T) {
+		image := "example.com/image:$@$(*"
+		_, err := ModifyImageRepository(image, testRepository)
+		g.Expect(err).To(MatchError(ContainSubstring("failed to parse image name")))
+	})
+}
+
+func TestModifyImageTag(t *testing.T) {
```
Comment (wfernandes) [inline util/container/image_test.go:176]: Feels like we can add more tests here. WDYT?

File: util/container/image.go [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+	normalisedTagName := SemverToOCIImageTag(tagName)
```
Comment (wfernandes) [inline util/container/image.go:108]: Should we be calling `ImageTagIsValid` in this function? Is any tag validation happening in the `reference` library?

File: util/container/image.go (hunk 18 above)
Comment (wfernandes) [reply util/container/image.go:107]: Yeah...my bad. I didn't understand at first what the output was but then I saw the tests and it made sense. Thanks! 🙂

File: util/container/image.go (hunk 14 above)
Comment (nader-ziada) [reply util/container/image.go:48]: ParseName does more validation

File: util/container/image.go (hunk 16 above)
Comment (nader-ziada) [reply util/container/image.go:46]: forgot to take that line out, nice catch

File: util/container/image.go (hunk 22 above)
Comment (nader-ziada) [reply util/container/image.go:108]: i think validation is happening in the reference library

File: util/container/image.go (hunk 14 above)
Comment (benmoss) [reply util/container/image.go:48]: Good catch. `ParseNormalizedNamed` will allow for "Docker hub"-style names, ala `busybox:latest`. `ParseNamed` will error `ErrNameNotCanonical` on those. I don't know if we want to support images like that.

Comment (nader-ziada) [reply util/container/image.go:48]: I found a few cases with checks to make sure  we don't get ErrNameNotCanonical so I assumed its better to use ParsedName

Comment (benmoss) [reply util/container/image.go:47]: I think it's better just that it forces people to be explicit, the Docker default registry nonsense is terrible

Comment (nader-ziada) [reply util/container/image.go:47]: agreed, I took the default out
